src/mlpack/methods/ann/layer/cross_entropy_error_with_logits_impl.hpp

```diff
@@ -0,0 +1,79 @@
+/**
+ * @file cross_entropy_error_with_logits_impl.hpp
+ * @author Kris Singh
+ *
+ * Implementation of the cross-entropy with logits performance function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_CROSS_ENTROPY_LOGIT_ERROR_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_CROSS_ENTROPY_LOGIT_ERROR_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "cross_entropy_error_with_logits.hpp"
+#include <mlpack/methods/ann/activation_functions/softplus_function.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+CrossEntropyErrorLogits<InputDataType, OutputDataType>
+::CrossEntropyErrorLogits()
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+double CrossEntropyErrorLogits<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, const arma::Mat<eT>&& target)
+{
+  eT loss = 0;
+  for (size_t i = 0; i < input.n_elem; i++)
+    if (input(i) > 0)
+    {
+      loss += input(i) - input(i) * target(i) +
+              SoftplusFunction::Fn(-std::abs(input(i)));
+    }
+    else
+    {
+      loss += input(i) * target(i) +
+              SoftplusFunction::Fn(-std::abs(input(i)));;
+    }
+
```

Comment:
Actually, you could simplify that. For example
```
arma::uvec positive = arma::find(input > 0);

loss = arma::accu(input(positive)) - arma::accu(input % target) + 
  arma::accu(SoftplusFunction::Fn(-arma::abs(input)));
```
However, it requires an extra memory allocation.